tracee-ebpf/tracee/tracee.bpf.c

```diff
@@ -2874,20 +2997,200 @@ int BPF_KPROBE(trace_security_socket_bind)
+{
+    local_net_id_t connect_id = {0};
+    u16 family = get_sock_family(sk);
+
+    if (family == AF_INET) {
+        net_conn_v4_t net_details = {};
+        get_network_details_from_sock_v4(sk, &net_details, 0);
+        if (net_details.local_port)
+            get_local_net_id_from_network_details_v4(sk, &connect_id, &net_details, family);
+    } else if (family == AF_INET6) {
+        net_conn_v6_t net_details = {};
+        get_network_details_from_sock_v6(sk, &net_details, 0);
+        if (net_details.local_port)
+            get_local_net_id_from_network_details_v6(sk, &connect_id, &net_details, family);
+    }
+
+    if (connect_id.port) {
+        if (tid != 0) {
+            net_ctx_t net_ctx;
+            net_ctx.host_tid = tid;
+            bpf_map_update_elem(&network_map, &connect_id, &net_ctx, BPF_ANY);
+        } else {
+            bpf_map_delete_elem(&network_map, &connect_id);
+        }
+    }
+
+    return 0;
+};
+
+SEC("kprobe/udp_sendmsg")
+int BPF_KPROBE(trace_udp_sendmsg)
+{
+    if (!should_trace())
+        return 0;
+
+    struct sock *sk = (struct sock *)PT_REGS_PARM1(ctx);
+    u32 tid = bpf_get_current_pid_tgid();
+
+    return net_map_update_or_delete_sock(sk, tid);
+};
+
+SEC("kprobe/__udp_disconnect")
+int BPF_KPROBE(trace_udp_disconnect)
+{
+    if (!should_trace())
+        return 0;
+
+    struct sock *sk = (struct sock *)PT_REGS_PARM1(ctx);
+
+    return net_map_update_or_delete_sock(sk, 0);
+};
+
+SEC("kprobe/udp_destroy_sock")
+int BPF_KPROBE(trace_udp_destroy_sock)
+{
+    if (!should_trace())
+        return 0;
+
+    struct sock *sk = (struct sock *)PT_REGS_PARM1(ctx);
+
+    return net_map_update_or_delete_sock(sk, 0);
+};
+
+SEC("kprobe/udpv6_destroy_sock")
+int BPF_KPROBE(trace_udpv6_destroy_sock)
+{
+    if (!should_trace())
+        return 0;
+
+    struct sock *sk = (struct sock *)PT_REGS_PARM1(ctx);
+
+    return net_map_update_or_delete_sock(sk, 0);
+};
+
+// include/trace/events/sock.h:
+// TP_PROTO(const struct sock *sk, const int oldstate, const int newstate)
+SEC("raw_tracepoint/inet_sock_set_state")
+int tracepoint__inet_sock_set_state(struct bpf_raw_tracepoint_args *ctx)
+{
+    local_net_id_t connect_id = {0};
+    net_ctx_ext_t net_ctx_ext = {0};
+
+    struct sock *sk = (struct sock *)ctx->args[0];
+    int new_state = ctx->args[2];
+
+    // Sometimes the socket state may be changed by other contexts that handle the tcp network stack (e.g. network driver).
+    // In these cases, we won't pass the should_trace() check.
+    // To overcome this problem, we save the socket pointer in sock_ctx_map in states that we observed to have the correct context.
+    // We can then check for the existence of a socket in the map, and continue if it was traced before.
+    net_ctx_ext_t *sock_ctx_p = bpf_map_lookup_elem(&sock_ctx_map, &sk);
+    if (!sock_ctx_p) {
+        if (!should_trace()) {
+            return 0;
+        }
+    }
+
```

Comment:
This logic seems VERY good for me (saving socket pointer, updating the net bpf map on ESTABLISHED, etc). I'm only concerned on socket being shared among different tasks and logic synchronization. ​

One corner case I can think of would be: tracee is starting and will trace only new pids(). Or specific pids will be traced, but not others. It gets to trace PID 1 of 3 pids. All sharing the same socket (due to fd passing). During its start it starts tracing PID 3 but not PIDs 1 and 2.

PID 1 would change socket state, trigger the tracepoint, socket would exist in sock_ctx_map because of PID 2 (but now we are coming to this trace point from PID 1). 

> This would succeed in setting the new port (from 0 to something):
> 
> bpf_map_update_elem(&sock_ctx_map, &sk, &net_ctx_ext, BPF_ANY);
>
> This would fail:
> 
> bpf_map_update_elem(&network_map, &connect_id, &net_ctx_ext, BPF_ANY);
>
> because of net_ctx_ext.host_tid being different, wouldn't it ?
>
> Would this to make this logic fail ?

Of course this is a corner case, I'm not even sure it is possible... but that raises the concern of atomicity of having 2 maps and different handles/keys to them with parallel executions of different entry points.

Reply:
We must have two different maps here. The explanation for this is simple:
1. The network map is used by the tc bpf program, where there is no process context, so the map key should only contain network related fields, and not process or socket related stuff
2. On the other hand, we want to save the context between two different tcp states, where either process context is "incorrect" or network context is incomplete (no local port).

So if we are using this tracepoint, I don't see any other choice.

Anyways, I don't see a problem with the use case you presented. The pid is not part of the key in any of the maps. So if a socket was inserted to the map, it will stay there until it is closed. Also, the network map don't care about the pid (which is part of the map value) as long as the key exists. One more thing, as we use BPF_ANY to update the map, it will be updated even if the key exist or not